File: 106/106.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder, postorder):
        index_map = {value: idx for idx, value in enumerate(inorder)}

        def build(in_left, in_right, post_left, post_right):
            if post_left > post_right:
                return None

            root_val = postorder[post_right]
            root = TreeNode(root_val)

            root_index = index_map[root_val]

            # Elements left of the root_index are part of the left subtree
            left_subtree_length = root_index - in_left
            root.left = build(in_left, root_index - 1, post_left, post_left + left_subtree_length - 1)
            root.right = build(root_index + 1, in_right, post_left + left_subtree_length, post_right - 1)

            return root

        return build(0, len(inorder) - 1, 0, len(postorder) - 1)

# Helper function to print tree in preorder traversal
def print_preorder(node):
    if node is None:
        return
    print(node.val, end=' ')
    print_preorder(node.left)
    print_preorder(node.right)

if __name__ == "__main__":
    inorder = [9, 3, 15, 20, 7]
    postorder = [9, 15, 7, 20, 3]
    solution = Solution()
    result_tree = solution.buildTree(inorder, postorder)

    print("Preorder of constructed tree:", end=' ')
    print_preorder(result_tree)
    print()

File: 106/106.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    unordered_map<int, int> indexMap;
    vector<int> postorder;

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        this->postorder = postorder;
        for (int i = 0; i < inorder.size(); ++i) {
            indexMap[inorder[i]] = i;
        }
        return build(0, inorder.size() - 1, 0, postorder.size() - 1);
    }

    TreeNode* build(int inLeft, int inRight, int postLeft, int postRight) {
        if (postLeft > postRight) {
            return nullptr;
        }
        
        int rootVal = postorder[postRight];
        TreeNode* root = new TreeNode(rootVal);
        int rootIndex = indexMap[rootVal];

        root->left = build(inLeft, rootIndex - 1, postLeft, postLeft + rootIndex - inLeft - 1);
        root->right = build(rootIndex + 1, inRight, postLeft + rootIndex - inLeft, postRight - 1);

        return root;
    }
};

// Helper function to print tree in preorder traversal
void printPreorder(TreeNode* node) {
    if (node == nullptr) {
        return;
    }
    cout << node->val << " ";
    printPreorder(node->left);
    printPreorder(node->right);
}

int main() {
    vector<int> inorder = {9,3,15,20,7};
    vector<int> postorder = {9,15,7,20,3};

    Solution solution;
    TreeNode* resultTree = solution.buildTree(inorder, postorder);

    cout << "Preorder of constructed tree: ";
    printPreorder(resultTree);
    cout << endl;

    return 0;
}
